language_service/src/compilation.rs

```diff
@@ -0,0 +1,258 @@
+    /// contains a single `Source`.
+    OpenDocument,
+    /// A Q# notebook. In a notebook compilation, the user package
+    /// contains multiple `Source`s, with each source corresponding
+    /// to a cell.
+    Notebook,
+}
+
+impl Compilation {
+    /// Creates a new `Compilation` by compiling source from a single open document.
+    pub(crate) fn new_open_document(
+        source_name: &str,
+        source_contents: &str,
+        package_type: PackageType,
+        target_profile: TargetProfile,
+    ) -> Self {
+        trace!("compiling document {source_name}");
+        // Source map only contains the current document.
+        let source_map = SourceMap::new([(source_name.into(), source_contents.into())], None);
+
+        let mut package_store = PackageStore::new(compile::core());
+        let std_package_id = package_store.insert(compile::std(&package_store, target_profile));
+
+        let (unit, errors) = compile::compile(
+            &package_store,
+            &[std_package_id],
+            source_map,
+            package_type,
+            target_profile,
+        );
+
+        let package_id = package_store.insert(unit);
+
+        Self {
+            package_store,
+            user: package_id,
+            errors,
+            kind: CompilationKind::OpenDocument,
+        }
+    }
+
+    /// Creates a new `Compilation` by compiling sources from notebook cells.
+    pub(crate) fn new_notebook<'a, I>(cells: I) -> Self
+    where
+        I: Iterator<Item = (&'a str, &'a str)>,
+    {
+        trace!("compiling notebook");
+        let mut compiler = Compiler::new(
+            true,
+            SourceMap::default(),
+            PackageType::Lib,
+            TargetProfile::Full,
+        )
+        .expect("expected incremental compiler creation to succeed");
+
+        let mut errors = Vec::new();
+        for (name, contents) in cells {
+            trace!("compiling cell {name}");
+            let increment = compiler
+                .compile_fragments(name, contents, |cell_errors| {
+                    errors.extend(cell_errors);
+                    Ok(()) // accumulate errors without failing
+                })
+                .expect("compile_fragments_acc_errors should not fail");
+
+            compiler.update(increment);
+        }
+
+        let (package_store, package_id) = compiler.into_package_store();
+
+        Self {
+            package_store,
+            user: package_id,
+            errors,
+            kind: CompilationKind::Notebook,
+        }
+    }
+
+    /// Gets the `CompileUnit` associated with user (non-library) code.
+    pub fn user_unit(&self) -> &CompileUnit {
+        self.package_store
+            .get(self.user)
+            .expect("expected to find user package")
+    }
+
+    /// Returns the package and `SourceMap` offset associated with a code location.
+    pub(crate) fn resolve_offset(&self, source_name: &str, offset: u32) -> (&AstPackage, u32) {
```

Comment:
Yeah I don't like how this refactoring came out. I took your naming suggestion, and dropped the `AstPackage` from the return value since it only ever returns the user package and the caller already has a reference to that. Thanks.

Reply:
Sorry I'm not seeing the change you mentioned. Is there a commit that hasn't been pushed yet?

Reply:
Looks like the PR got out of sync with your branch somehow. Maybe it is just a matter of waiting for it to get back in sync.

Reply:
I see the sync fixed the confusion. Thanks for looking into this!